pkg/cvo/updatepayload.go

```diff
@@ -217,123 +219,188 @@ func (r *payloadRetriever) fetchUpdatePayloadToDir(ctx context.Context, dir stri
+				return kubeClient.CoreV1().Pods(namespace).Watch(ctx, metav1.ListOptions{FieldSelector: fieldSelector})
+			},
+		},
+		&corev1.Pod{},
+		nil,
+		func(event watch.Event) (bool, error) {
+			p, ok := event.Object.(*corev1.Pod)
+			if !ok {
+				klog.Errorf("expecting Pod but received event with kind: %s", event.Object.GetObjectKind())
+				return false, fmt.Errorf("expecting Pod but received event with kind: %s", event.Object.GetObjectKind())
+			}
+			switch phase := p.Status.Phase; phase {
+			case corev1.PodPending:
+				klog.Infof("Pod %s is pending", name)
+				// This is a workaround to collect the failures when the pod is still pending such as "reason: SignatureValidationFailed"
+				// before https://github.com/kubernetes/kubernetes/pull/127918 lands into production
+				// The message looks like 'image pull failed for quay.io/openshift-release-dev/ocp-release@sha256:digest because the signature validation failed: Source image rejected: A signature was required, but no signature exists'
+				if statuses := collectStatuses(p.Status, func(msg string) bool { return strings.Contains(strings.ToLower(msg), "fail") }); len(statuses) > 0 {
+					klog.Errorf("Pod %s failed at pending with reason %q and message %q and status %s", name, p.Status.Reason, p.Status.Message, strings.Join(statuses, ","))
+					return false, fmt.Errorf("pod %s failed at pending with reason %q and message %q and status %s", name, p.Status.Reason, p.Status.Message, strings.Join(statuses, ","))
+				}
+				return false, nil
+			case corev1.PodRunning:
+				klog.Infof("Pod %s is running, waiting for its completion ...", name)
+				return false, nil
+			case corev1.PodSucceeded:
+				klog.Infof("Pod %s succeeded", name)
+				return true, nil
+			case corev1.PodFailed:
+				// The reason "DeadlineExceeded" and the message "Pod was active on the node longer than the specified deadline"
+				// from the pod's status are not very useful for debugging, so we collect containers status with non-empty message as well
+				statuses := collectStatuses(p.Status, func(msg string) bool { return msg != "" })
+				klog.Errorf("Pod %s failed with reason %q and message %q and status %s", name, p.Status.Reason, p.Status.Message, strings.Join(statuses, ","))
+				return false, fmt.Errorf("pod %s failed with reason %q and message %q and status %s", name, p.Status.Reason, p.Status.Message, strings.Join(statuses, ","))
+			default:
+				klog.Errorf("Pod %s is with unexpected phase %s", name, phase)
+				return false, fmt.Errorf("pod %s is with unexpected phase %s", name, phase)
+			}
+		},
+	)
+	return err
 }
 
-// pruneJobs deletes the older, finished jobs in the namespace.
-// retain - the number of newest jobs to keep.
-func (r *payloadRetriever) pruneJobs(ctx context.Context, retain int) error {
+// prunePods deletes the older, finished pods in the namespace.
+func (r *payloadRetriever) prunePods(ctx context.Context) error {
```

Comment:
Still quite a lot of `r.kubeClient.BatchV1().Jobs` usage - seems like this needs to be ported  too?

Reply:
we need to preserve some Job pruning for a transitional period, to handle history like:

1. Job-based CVO creates Jobs to download update targets.
2. Update to a Pod-based CVO but leak a Job somehow.
3. New Pod-based CVO still knows how to reap Jobs, and cleans them out.

If 4.19 is the first with Pod-based CVOs, 4.20 could drop the Job reaping.  But 4.19 would have both Pod- and Job-reaping.

Reply:
Perhaps Petr wants to do Job pruning out of the function `prunePods`?

Reply:
I think I just needed the transitional period called out more loudly in the code. There are comments but they have not clicked for me.